src/components/tl/cuda/tl_cuda_topo.c

```diff
@@ -274,19 +285,38 @@ ucc_status_t ucc_tl_cuda_topo_num_links(const ucc_tl_cuda_topo_t *topo,
                                         int *num_links)
 {
     ucc_status_t status;
-    ucc_tl_cuda_topo_node_t *dev1_node;
-    ucc_tl_cuda_topo_link_t *link;
+    ucc_tl_cuda_topo_node_t *dev1_node, *dev2_node;
+    ucc_tl_cuda_topo_node_t *link_node;
+    ucc_tl_cuda_topo_link_t *dev1_link, *dev2_link;
 
     *num_links = 0;
     status = ucc_tl_cuda_topo_graph_find_by_id(topo, dev1, &dev1_node);
     if (status != UCC_OK) {
         return status;
     }
+    status = ucc_tl_cuda_topo_graph_find_by_id(topo, dev2, &dev2_node);
+    if (status != UCC_OK) {
+        return status;
+    }
 
-    ucc_list_for_each(link, &dev1_node->link.list_link, list_link) {
-        if (ucc_tl_cuda_topo_device_id_equal(&link->pci_id, dev2)) {
-            *num_links += link->width;
-            return UCC_OK;
+    ucc_list_for_each(dev1_link, &dev1_node->link.list_link, list_link) {
+        if (ucc_tl_cuda_topo_device_id_equal(&dev1_link->pci_id, dev2)) {
+            *num_links += dev1_link->width;
+        } else {
+            /* check if peer device is switch and it connects given devices */
+            status = ucc_tl_cuda_topo_graph_find_by_id(topo, &dev1_link->pci_id,
+                                                       &link_node);
+            if (status != UCC_OK) {
+                return status;
+            }
+            if (link_node->type == UCC_TL_CUDA_TOPO_DEV_TYPE_SWITCH) {
+                ucc_list_for_each(dev2_link, &dev2_node->link.list_link, list_link) {
+                    if (ucc_tl_cuda_topo_device_id_equal(&dev2_link->pci_id,
```

Comment:
is switch have only one PCI_ID? or each target device have a PCI ID?

Reply:
Each switch has one PCI_ID, but two GPU devices might be connected through multiple NVSwitches. For example If you look at DGX A100 topology (https://docs.nvidia.com/dgx/pdf/dgxa100-user-guide.pdf  1.3 DGX A100 System Topology page  13) there are 6 NVSwitches and each GPU has two connections to each switch.